server/src/main/java/org/elasticsearch/action/search/SearchResponse.java

```diff
@@ -204,8 +204,6 @@ public int getSkippedShards() {
      * The failed number of shards the search was executed on.
      */
     public int getFailedShards() {
-        // we don't return totalShards - successfulShards, we don't count "no shards available" as a failed shard, just don't
```

Comment:
We use the `SearchResponse` as partial response in `_async_search` so that's not possible.
This change is useful though because it clarifies that `shardFailures.length + successfullShards != totalShards` represents a partial response where some shard requests are still in-flight.

Reply:
Ah, ok. Perhaps we can add the assertion at the callsites creating the full `SearchResponse` instead? Like `AbstractSearchAsyncAction.buildSearchResponse`.

Reply:
Added in https://github.com/elastic/elasticsearch/pull/64337/commits/235416573feb3cae1b85b52417a6f1eaa6ce2edf